### PR TITLE
Fix duplicate parquet output in `to_parquet`

### LINKED ISSUE #6445: to_parquet with pyarrow has produced duplicated files, after a worker has crashed
**What happened**:
Our job was writing 100 partitions, using the pyarrow engine. One of the workers crashed, after it has written a parquet file. As a consequence, we had 101 parquet files with duplicate data in our s3 bucket.

We observed, that the `to_parquet` task got re-scheduled and re-executed on an other worker, as it was in `memory` and waiting to be processed by the `write_metadata` task.

**What you expected to happen**:
100 parquet partitions in our s3 bucket, even when a worker dies after writing.

**Minimal Complete Verifiable Example**:
```python
import pandas as pd
from dask import dataframe

df = pd.DataFrame(["A", "B"], columns=["part"])
df['data'] = 1
ddf = dataframe.from_pandas(df, npartitions=1)

for i in range(3):
    ddf.to_parquet("out/fastparquet", engine="fastparquet", partition_on="part")
    ddf.to_parquet("out/pyarrow", engine="pyarrow", partition_on="part")
```

produces the following output

```
out
├── fastparquet
│   ├── _common_metadata
│   ├── _metadata
│   ├── part=A
│   │   └── part.0.parquet
│   └── part=B
│       └── part.0.parquet
└── pyarrow
    ├── _common_metadata
    ├── _metadata
    ├── part=A
    │   ├── 33a45f31937e467880cd24d11c055133.parquet
    │   ├── 691d5e18d5b745fe8cbb203cbf4f7523.parquet
    │   └── aa4a303bbaec44c1ae15c92cbcb1dfad.parquet
    └── part=B
        ├── 6794a35ed580479d90087c4b13d41442.parquet
        ├── d0847e218a9142429014300bfb7f1ee9.parquet
        └── f526b99981a34595a9bbec220e3d21e4.parquet
```

### PR DESCRIPTION
**Pull request by michaelnarodovitch**
May fix #6445 .

Not sure, if the `guid` is required for something.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by jsignell** [comment]
Thanks @michaelnarodovitch for submitting this fix. This seems reasonable to me. Pinging @dask/io for a review.

**Comment by martindurant** [comment]
Would you mind being specific about what you are changing here, @michaelnarodovitch , what was duplicated? 
I don't quite follow from glancing at the code (I don't immediately know what the guid was for).

**Comment by jsignell** [comment]
@martindurant - there is more context in the issue.

**Comment by rjzamora** [comment]
> Would you mind being specific about what you are changing here, @michaelnarodovitch , what was duplicated?
I don't quite follow from glancing at the code (I don't immediately know what the guid was for).

The `_write_partitioned` logic was borrowed from pyarrow, which uses a random uuid label for the files within a partition directory.  It seems that the random nature of the file name may have caused the same data to be written to two separate files when one of the distributed workers died.  When the file name is deterministic, the rescheduled task will overwrite the data instead of duplicating it.

It makes sense to use guid to avoid name collisions ion general.  However, it shouldn't be necessary to use a random name in dask, since each partition is already given a unique file name.  I'm still pondering this a bit, but I'm pretty sure this PR makes sense.

File: dask/dataframe/io/tests/test_parquet.py [hunk 1]
```diff
@@ -1089,6 +1089,25 @@ def test_partition_on(tmpdir, engine):
...
+        }
+    )
+    d = dd.from_pandas(df, npartitions=2)
+
+    for _ in range(2):
+        d.to_parquet(tmpdir, partition_on=["a1", "a2"], engine=engine)
+
+    out = dd.read_parquet(tmpdir, engine=engine).compute()
+
+    assert len(df) == len(out)
```
Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py:1108]: Please add a commend referencing your issue, as this assert is what would have failed in the previous version.

**Comment by martindurant** [comment]
Maybe the test should check the explicit name of at least one output file, since we are asserting it is deterministic and we know what it ought to be. Do you also get something like part.0.parquet now, as for fastparquet? (where the number corresponds to which original partition is contains data from) That would be nice consistency.

File: dask/dataframe/io/tests/test_parquet.py (hunk 1 above)
Comment (michaelnarodovitch) [reply dask/dataframe/io/tests/test_parquet.py:1106]: Added a reference.

**Comment by jsignell** [comment]
@michaelnarodovitch you seem to have brought a lot of style changes with that latest commit that I suspect were unintentional. You might want to reset back one commit and check that you don't have any default formatters in your editor.

**Comment by michaelnarodovitch** [comment]
Thanks to everybody for having a look and the great feedback.

As far as I am concerned, the present test guarantees, that the issue doesn't occur in future.

With the present changes, filenames for both, `pyarrow` and `fastparquet` will match `part.*.parquet`. Is it something we would like to test/enforce? IMO it is beneficial, if filenames don't change accidentally. Happy to add that.

Leaving to vacation right now and will follow-up in the upcoming week. Just pumped out the last commit :( Will roll it back soon!

**Comment by michaelnarodovitch** [comment]
> Maybe the test should check the explicit name of at least one output file, since we are asserting it is deterministic and we know what it ought to be. Do you also get something like part.0.parquet now, as for fastparquet? (where the number corresponds to which original partition is contains data from) That would be nice consistency.

Added assertions for the filenames to the test.

I found, that there is already some assertions in https://github.com/michaelnarodovitch/dask/blob/fix/dask/dataframe/io/tests/test_parquet.py#L166

These assertions don't use the `partition_on` flag, though.

**Comment by michaelnarodovitch** [comment]
As far as I am concerned, this PR is ready again. We would be really happy to see that in production :)

**Comment by jsignell** [comment]
This looks good to me and will have the side-effect of getting rid of some pesky warnings. Thanks @michaelnarodovitch for finishing this up and congrats on your first dask PR!

**Pull request merged**